### PR TITLE
Upgrade/thin skinny em

### PR DESCRIPTION
**Pull request by montdidier**
Update to thin 1.8 (will compile under Xcode 12) and event machine 1.2 

Some of the changes to make this possible include. 

- Moving `Process.daemon` to before the the eventmachine loop
- Handle signals in mailcatcher (changes in thin made this necessary)
- Removed the explicit gem activations which I don't think are necessary (anymore?)
- Removed the EventMachine monkey patch which should not be necessary anymore
- Added some constraints on dependencies

Depends on https://github.com/sj26/skinny/pull/8

Should supersede https://github.com/sj26/mailcatcher/pull/431

### CODE REVIEW

**Comment by sj26** [comment]
Neat! What happens if you start mailcatcher and the port is in use?

Do websockets work? Last time I tried this it broke.

**Comment by montdidier** [comment]
The only issue I have noticed is the console output is a bit messy - because control has returned to the shell. 

If the port is in use it still print outs because it has not yet point the std io to `/dev/null`

**Comment by montdidier** [comment]
The other option might be to test the ports ahead of time using another API of some kind. Unfortunately it seems like the Servers need to be called within the run loop. I'm happy to be wrong about that though.

**Comment by sj26** [comment]
Yeah I'd love to open the socket, handle failure, or hand the open IO to eventmachine after the fork, but I couldn't figure that quickly last time I tried.

**Comment by tagliala** [comment]
Hi, is there anything that I can do to move this forward?

At the moment I'm using `rvm ruby-2.4.9 do mailcatcher -f -v`, but it would be nice to run mailcatcher on supported Ruby versions

**Comment by montdidier** [comment]
I don't think there is anything super wrong with the way it currently is apart from the occasional console output formatting issue on startup.

I did look into fixing it the way suggested but it is a fair whack more work because it looks like it I would need to change the C part of eventmachine. I am C fluent but I haven't yet built up the momentum to tackle it.

File: Gemfile [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 source "https://rubygems.org"
 
+gem "skinny", github: "montdidier/skinny", branch: "upgrade/eventmachine-thin"
```
Comment (montdidier) [inline Gemfile:5]: This will not be necessary once the new skinny gem is published.

**Comment by montdidier** [comment]
> I did look into fixing it the way suggested but it is a fair whack more work because it looks like I would need to change the C part of eventmachine. I am C fluent but I haven't yet built up the momentum to tackle it.

Actually I think I lied. I just spied a way to do it. I'll update this PR soon.

File: lib/mail_catcher/tcpserver.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+module MailCatcher
+  class TcpServer < Thin::Backends::Base
+    # Address and port on which the server is listening for connections.
+    attr_accessor :socket
```
Comment (montdidier) [inline lib/mail_catcher/tcpserver.rb:4]: probably can remove this. I don't use it.

Comment (montdidier) [reply lib/mail_catcher/tcpserver.rb:4]:
```suggestion
```

**Comment by montdidier** [comment]
> Yeah I'd love to open the socket, handle failure, or hand the open IO to eventmachine after the fork, but I couldn't figure that quickly last time I tried.

I think this has been achieved.

**Comment by sj26** [comment]
Thank you for this, but we're closing in favour of moving away from eventmachine, thin and skinny in #467.

**Pull request closed without merging**